src/Lean/Server/Rpc/Basic.lean

```diff
@@ -15,98 +16,114 @@ For example, the client can format an `Expr` without transporting the whole `Env
-  rpcEncode v := match v with
-    | none => pure none
-    | some v => some <$> rpcEncode v
-  rpcDecode v := match v with
-    | none => pure none
-    | some v => some <$> rpcDecode v
-
--- TODO(WN): instance [RpcEncoding α β] [Traversable t] : RpcEncoding (t α) (t β)
-instance [RpcEncoding α β] : RpcEncoding (Array α) (Array β) where
-  rpcEncode a := a.mapM rpcEncode
-  rpcDecode b := b.mapM rpcDecode
-
-instance [RpcEncoding α α'] [RpcEncoding β β'] : RpcEncoding (α × β) (α' × β') where
-  rpcEncode := fun (a, b) => do
-    let a' ← rpcEncode a
-    let b' ← rpcEncode b
-    return (a', b')
-  rpcDecode := fun (a', b') => do
-    let a ← rpcDecode a'
-    let b ← rpcDecode b'
-    return (a, b)
-
-structure RpcEncoding.DerivingParams where
-  withRef : Bool := false
+class RpcEncodable (α : Type) where
+  rpcEncode : α → StateM RpcObjectStore Json
+  rpcDecode : Json → ExceptT String (ReaderT RpcObjectStore Id) α
+export RpcEncodable (rpcEncode rpcDecode)
+
+instance : Nonempty (RpcEncodable α) :=
+  ⟨{ rpcEncode := default, rpcDecode := default }⟩
+
+instance [FromJson α] [ToJson α] : RpcEncodable α where
+  rpcEncode a := return toJson a
+  rpcDecode j := ofExcept (fromJson? j)
+
+instance [RpcEncodable α] : RpcEncodable (Option α) where
+  rpcEncode v := toJson <$> v.mapM rpcEncode
+  rpcDecode j := do Option.mapM rpcDecode (← fromJson? j)
+
+-- TODO(WN): instance [RpcEncodable α β] [Traversable t] : RpcEncodable (t α) (t β)
+
+instance [RpcEncodable α] : RpcEncodable (Array α) where
+  rpcEncode a := toJson <$> a.mapM rpcEncode
+  rpcDecode b := do Array.mapM rpcDecode (← fromJson? b)
+
+instance [RpcEncodable α] [RpcEncodable β] : RpcEncodable (α × β) where
+  rpcEncode := fun (a, b) => return toJson (← rpcEncode a, ← rpcEncode b)
+  rpcDecode j := do
+    let (a, b) ← fromJson? j
+    return (← rpcDecode a, ← rpcDecode b)
 
 /-- Marks fields to encode as opaque references in LSP packets. -/
 structure WithRpcRef (α : Type u) where
   val : α
   deriving Inhabited
 
-namespace WithRpcRef
-
-variable {m : Type → Type} [Monad m] [MonadRpcSession m]
-
-/-- This is unsafe because we must ensure that:
-- the stored `NonScalar` is never used to access the value as a type other than `α`
-- the type `α` is not a scalar -/
-protected unsafe def encodeUnsafe [Monad m] (typeName : Name) (r : WithRpcRef α) : m Lsp.RpcRef := do
-  let obj := @unsafeCast α NonScalar r.val
-  rpcStoreRef typeName obj
-
-protected unsafe def decodeUnsafeAs (α) (typeName : Name) (r : Lsp.RpcRef) : ExceptT String m (WithRpcRef α) := do
-  match (← rpcGetRef r) with
-    | none => throw s!"RPC reference '{r}' is not valid"
-    | some (nm, obj) =>
-      if nm != typeName then
-        throw s!"RPC call type mismatch in reference '{r}'\nexpected '{typeName}', got '{nm}'"
-      return WithRpcRef.mk <| @unsafeCast NonScalar α obj
+instance [Dynamic α] : RpcEncodable (WithRpcRef α) :=
+  { rpcEncode, rpcDecode }
+where
+  -- separate definitions to prevent inlining
```

Comment:
Why is inlining undesirable here? Just code size?

Reply:
Yes, this is mostly code size (and being able to read the output of `set_option trace.compiler.ir.result true`).